### ISSUE TITLE
Broken build, likely outdated code in PSM2 component

### ISSUE THREAD

**Issue by dalcinl** [open-mpi/ompi#10849]
This is with current main branch, using an out-of-source build. I configured this way:

```
$ ../ompi/configure --prefix=/home/devel/mpi/openmpi/dev --without-ofi --without-ucx --with-pmix=internal --with-prrte=internal --with-libevent=internal --with-hwloc=internal --enable-debug --enable-mem-debug --disable-man-pages --disable-sphinx
```

Note that I have PSM2 installed:

```
$ rpm -qa | grep psm2
libpsm2-11.2.206-2.fc36.x86_64
libpsm2-devel-11.2.206-2.fc36.x86_64
```

The build fails at the point of building the PSM2 component

```
$ make
...
Making all in mca/mtl/psm2
make[2]: Entering directory '/home/dalcinl/Devel/REPOS/ompi-BUILD/ompi/mca/mtl/psm2'
  CC       mtl_psm2.lo
In file included from ../../../../../ompi/ompi/mca/mtl/psm2/mtl_psm2.c:32:
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h: In function 'ompi_mtl_datatype_pack':
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h:71:93: error: macro "malloc" passed 3 arguments, but takes just 1
   71 |             opal_accelerator.malloc(MCA_ACCELERATOR_NO_DEVICE_ID, &iov.iov_base, *buffer_len);
      |                                                                                             ^
In file included from ../../../../opal/include/opal_config.h:2408,
                 from ../../../../../ompi/ompi/include/ompi_config.h:29,
                 from ../../../../../ompi/ompi/mca/mtl/psm2/mtl_psm2.c:28:
../../../../../ompi/opal/include/opal_config_bottom.h:339: note: macro "malloc" defined here
  339 | #        define malloc(size) opal_malloc((size), __FILE__, __LINE__)
      | 
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h: In function 'ompi_mtl_datatype_recv_buf':
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h:106:86: error: macro "malloc" passed 3 arguments, but takes just 1
  106 |             opal_accelerator.malloc(MCA_ACCELERATOR_NO_DEVICE_ID, buffer, *buffer_len);
      |                                                                                      ^
../../../../../ompi/opal/include/opal_config_bottom.h:339: note: macro "malloc" defined here
  339 | #        define malloc(size) opal_malloc((size), __FILE__, __LINE__)
      | 
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h: In function 'ompi_mtl_datatype_unpack':
../../../../../ompi/ompi/mca/mtl/base/mtl_base_datatype.h:135:71: error: macro "free" passed 2 arguments, but takes just 1
  135 |             opal_accelerator.free(MCA_ACCELERATOR_NO_DEVICE_ID, buffer);
      |                                                                       ^
../../../../../ompi/opal/include/opal_config_bottom.h:351: note: macro "free" defined here
  351 | #        define free(ptr) opal_free((ptr), __FILE__, __LINE__)
      | 
make[2]: *** [Makefile:1556: mtl_psm2.lo] Error 1
make[2]: Leaving directory '/home/dalcinl/Devel/REPOS/ompi-BUILD/ompi/mca/mtl/psm2'
make[1]: *** [Makefile:2781: all-recursive] Error 1
make[1]: Leaving directory '/home/dalcinl/Devel/REPOS/ompi-BUILD/ompi'
make: *** [Makefile:1515: all-recursive] Error 1
```

Of course, adding `--without-psm2` to configure fixes my build.

**Comment by jsquyres** [comment]
@wckzhang I'm guessing that this was as result of merging #10069.

@BrendanCunningham Can you guys look at this?

**Comment by BrendanCunningham** [comment]
@jsquyres Yep, assign it to me.

**Comment by wckzhang** [comment]
Yeah this was a result of the accelerator framework, also needs to update the free after code in ompi mtl datatype pack to also do accelerator frees.

**Comment by wckzhang** [comment]
Looked into it, this is actually because with --enable-mem-debug, malloc and free have been macro'd to opal_malloc and opal_free. The accelerator API uses malloc and free as its name, I'll change that to mem_alloc and mem_release

**Comment by wckzhang** [comment]
PR to fix: https://github.com/open-mpi/ompi/pull/10856/commits

**Comment by wckzhang** [comment]
@dalcinl it doesn't need to be backported yet as the original patch wasn't merged into 5.0.x, it can be closed. I'll make sure to backport all the necessary patches at once
